phoenix-core/src/main/java/org/apache/phoenix/iterate/SpoolingByteBufferSegmentQueue.java

```diff
@@ -0,0 +1,357 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.iterate;
+
+import org.apache.commons.io.input.CountingInputStream;
+import org.apache.commons.io.output.DeferredFileOutputStream;
+
+import org.apache.phoenix.memory.MemoryManager;
+import org.apache.phoenix.memory.MemoryManager.MemoryChunk;
+import org.apache.phoenix.query.QueryServices;
+import org.apache.phoenix.query.QueryServicesOptions;
+
+
+import java.io.*;
+import java.util.AbstractQueue;
+import java.util.Iterator;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.phoenix.monitoring.GlobalClientMetrics.*;
+
+
+public abstract class SpoolingByteBufferSegmentQueue<T>  extends AbstractQueue<T> {
+
+    private ResultQueue<T> spoolFrom;
+
+    private boolean closed ;
+    private boolean flushed;
+    private DeferredFileOutputStream spoolTo;
+    private MemoryChunk chunk;
+    private int size = 0;
+    private long inMemByteSize = 0L;
+    private int index;
+
+
+
+    SpoolingByteBufferSegmentQueue(int index, MemoryManager mm, final int thresholdBytes, String spoolDirectory)  {
+
+        long startTime = System.currentTimeMillis();
+        chunk  = mm.allocate(0, thresholdBytes);
+        long waitTime = System.currentTimeMillis() - startTime;
+        GLOBAL_MEMORY_WAIT_TIME.update(waitTime);
+
+        int size = (int)chunk.getSize();
+        spoolTo = new DeferredFileOutputStream(size, "ResultSpooler",".bin", new File(spoolDirectory)) {
+            @Override
+            protected void thresholdReached() throws IOException {
+                try {
+                    super.thresholdReached();
+                } finally {
+                    chunk.close();
+                }
+            }
+        };
+
+
+    }
+
+    public int index() {
+        return this.index;
+    }
+
+
+
+    protected abstract InMemoryResultQueue<T> createInMemoryResultQueue(byte[] bytes, MemoryChunk memoryChunk);
+
+    protected abstract OnDiskResultQueue<T> createOnDiskResultQueue(File file);
+
+    @Override
+    public boolean offer(T t) {
+        if (closed || flushed){
+            return false;
+        }
+        boolean result = writeRecord(t, spoolTo);
+        if(result){
```

Comment:
I'm pretty lost at why this "if" block here, means we flush to disk all the time? I'll stop here though, leaving all other doubts, coz this all (XXXQueue classes) is gonna take a rewrite I think.